UDF Support: Direct calls and GCS hosted code

gramster opened the pull request:
Add the ability to call UDFs directly in the SQL without having to compose them via Python code into a FunctionCall first, as well as reference code in GCS.

Issue #606

gramster commented:
It might be good to get this in before you leave if you have time as I think it is orthogonal to other changes suggested (GCS hosted UDFs, not using JS for testing).

gramster commented:
So I've realized I overlooked something here because of the example we have been using with Shakespeare. When we have the intermediate Python code to create a UDF 'call' we wrap the UDF in a 'SELECT field1, .... FROM fn(...)' fragment. However, we are not doing any substitution if we call the UDF directly in a %%sql module; we just note that we need to include the code when we issue the request. That means we run foul of the inability to do SELECT * from a UDF. We actually need to find the start and end parentheses that wrap the UDF call so that we can wrap the whole thing in a SELECT field1,... etc, which brings us back to parsing SQL. Sigh.

gramster commented:
The direct-call UDFs are working properly now with 'SELECT *' . I'm going to add some more tests but things are looking good.

nikhilk commented on sources/lib/api/gcp/bigquery/_udf.py:
```diff
@@ -13,34 +13,27 @@
+import gcp.storage
 
 
 class FunctionCall(object):
```
Do we still need this as well as FunctionEvaluation?
I am assuming we're no longer going to allow you to locally call a UDF directly.

gramster replied:
We want to keep FunctionEvaluation for local testing (I don't think we want to get rid of that; even the BQ web console uses local evaluation in the browser for UDF testing). But we don't really need FunctionCall if we expect users to only test locally or call the UDF within SQL.

nikhilk replied:
I think we only need to support calling UDFs within SQL.

Calling into the javascript to test locally is interesting, but I don't know how realistic, given the javascript you are writing can depend on javascript in GCS.

In fact, if you are trying to write scalar UDFs, the best way to do it right now is with the UDF is GCS as a reusable script, and the specific UDF for the specific query schema as inline. For example, I want to write a UDF for geohashing over lat+lon which I can write and store in GCS to be used by a query-specific UDF that operates over columns a,b,c,lat,lon and roundtrips the a,b,c part.

I am assuming testing with the actual BQ dremel engine is cheap enough over sample data, so why would be not always do that instead of testing in the browser... just seems like a cheap hack ... yes, I wrote it, and it was actually written as a demo over lunch at an all day offsite when someone asked me if notebooks could enable test scenarios :)

gramster replied:
I'll remove it, but once again though I feel I should point out that in-browser testing is what the BQ team themselves promote. So if it is a cheap hack its a cheap hack that the company promotes.

http://storage.googleapis.com/bigquery-udf-test-tool/testtool.html

gramster replied:
Actually, if we're supporting calling UDFs in only SQL then we don't need UDF.\_\_call\_\_, nor do we need FunctionEvaluation or FunctionCall; they can all go.

gramster replied:
Done

nikhilk commented on sources/lib/api/gcp/bigquery/_udf.py:
```diff
@@ -120,19 +84,77 @@ def __init__(self, inputs, outputs, name, implementation):
+    self._imports = imports
+    self._code = UDF._build_js(inputs, outputs, name, implementation, support_code)
+
+  def expand_imported_code(self):
```
Not sure we should keep pushing the browser running javascript as a way to test UDFs.

gramster replied:
It's a low cost option. BQ console does this too as it doesn't cost anything. It may be easier to debug as well (getting local console errors rather than errors back from BQ). So I'd argue for keeping it.

nikhilk replied:
So would we load everything from GCS as well? How do you know the dependencies of a single javascript function itself?

gramster replied:
Moot point now as this is gone.

nikhilk commented on sources/lib/api/gcp/bigquery/_udf.py:
```diff
@@ -120,19 +84,77 @@ def __init__(self, inputs, outputs, name, implementation):
+    # Build the JS from the individual bits with proper escaping of the implementation
+    if support_code is None:
+      support_code = ''
+    return '%s\n%s=%s;\nbigquery.defineFunction(\'%s\', %s, %s, %s);' % \
```
I think we should shift to named args... if not in this change, and sometime soon. These quickly become unreadable.

gramster replied:
I'm not sure what you mean; this is JS code.

nikhilk replied:
The string formatting is happening in python. Using named placeholders instead of %s would make this readable/understandable.

gramster replied:
Done

nikhilk commented on sources/lib/api/gcp/data/__init__.py:
```diff
@@ -27,10 +27,109 @@ def sql(sql_template, **kwargs):
+
+
+def tokenize(sql):
+  """ This is a crude tokenizer for our limited purposes.
```
Was this just for extracting out UDF uses? Maybe you mentioned, and I don't remember, but given a list of udf names, can we just look for "name(...)"?

So rather than finding all function calls, we have a fixed list of function calls to find.

gramster replied:
We need to find the corresponding closing parenthesis for any particular call, so we can wrap the whole thing in (SELECT f1, ... FROM ...). This tokenizer was created primarily for that, but once we were using it for that it made sense to use it to find the function calls as well; the resulting code is quite a bit simpler than an earlier version that used REs to find the calls and the tokenizer to find the call ends.

nikhilk replied:
Side note -- this shouldn't be public or perhaps even in __init__.py.

Secondly coroutines would be simpler to read through rather than hand-coded state machines, assuming they are an option. See http://eli.thegreenplace.net/2009/08/29/co-routines-as-an-alternative-to-state-machines/

I guess we're going to invariably end up in sql parsing. We should think about how we do this correctly across the scenarios -- parsing out UDF references, parsing begin/end of queries, comments in %%sql modules etc.

gramster replied:
Right now it is public as it is defined in gcp.data but used in gcp.bigquery. I can move it to gcp.bigquery although we may want it later.

gramster replied:
I have an alternative implementation; do you prefer this:

def _next_token(sql):
      i = 0
    
      # We use some lambda's to make the logic more clear. The start_* functions return
      # true if i is the index of the start of that construct, while the end_* functions
      # return true if i point to the first character beyond that construct or the end of the
      # content.
      #
      # We don't currently need numbers so the tokenizer here just does sequences of
      # digits as a convenience to shrink the total number of tokens. If we needed numbers
      # later we would need a special handler for these much like strings.
    
      start_multi_line_comment = lambda s, i: s[i] == '/' and i < len(s) - 1 and s[i + 1] == '*'
      end_multi_line_comment = lambda s, i: s[i - 2] == '*' and s[i - 1] == '/'
      start_single_line_comment = lambda s, i: s[i] == '-' and i < len(s) - 1 and s[i + 1] == '-'
      end_single_line_comment = lambda s, i: s[i - 1] == '\n'
      start_whitespace = lambda s, i: s[i].isspace()
      end_whitespace = lambda s, i: not s[i].isspace()
      start_number = lambda s, i: s[i].isdigit()
      end_number = lambda s, i: not s[i].isdigit()
      start_identifier = lambda s, i: s[i].isalpha() or s[i] == '_' or s[i] == '$'
      end_identifier = lambda s, i: not(s[i].isalnum() or s[i] == '_')
      start_string = lambda s, i: s[i] =='"' or s[i] == "'"
      always_true = lambda s, i: True
    
      while i < len(sql):
        start = i
        if start_multi_line_comment(sql, i):
          i += 2
          end_checker = end_multi_line_comment
        elif start_single_line_comment(sql, i):
          i += 2
          end_checker = end_single_line_comment
        elif start_whitespace(sql, i):
          i += 1
          end_checker = end_whitespace
        elif start_identifier(sql, i):
          i += 1
          end_checker = end_identifier
        elif start_number(sql, i):
          i += 1
          end_checker = end_number
        elif start_string(sql, i):
          # Special handling here as we need to check for escaped closing quotes.
          quote = sql[i]
          end_checker = always_true
          i += 1
          while i < len(sql) and sql[i] != quote:
            i += 2 if sql[i] == '\\' else 1
          i += 1  # skip quote
        else:
          # We return single characters for everything else
          i += 1
          end_checker = always_true
    
        while i < len(sql) and not end_checker(sql, i):
          i += 1
    
        (yield sql[start:i])

gramster commented:
I have removed all the \_\_call\_\_ related code, and changed the tokenizer to not be an FSM but a generator. Also in changing to the tokenizer it turned out we lost the ability to report on unresolvable UDF  names so that has been fixed; you'll see what looks like a big diff in Query constructor but its mostly an indentation change; git diff is pretty lame in that respect.

nikhilk commented on sources/lib/api/gcp/bigquery/_query.py:
```diff
@@ -66,23 +68,74 @@ def __init__(self, sql, scripts=None, context=None, values=None, **kwargs):
+
+      if token not in udf_dict:
+        utoken = token.upper()
+        if utoken != 'TABLE_DATE_RANGE' and utoken != 'TABLE_DATE_RANGE_STRICT' and \
```
Will it work if we don't check for this? By work, I mean BigQuery will raise an appropriate error message. It would be better if we didn't hardcode some select function names here.

gramster replied:
Do you mean should we not raise errors ourselves if UDFs are not defined? If so, yes, BigQuery will report it, although it will uppercase the UDF name. E.g.:

    RequestException: Unknown TVF: FILTER

for a call to 'filter'.

nikhilk commented on sources/lib/api/gcp/bigquery/_udf.py:
```diff
@@ -89,50 +73,13 @@ def _build_js(inputs, outputs, name, implementation):
+            'bigquery.defineFunction(\'{name}\', {inputs}, {outputs}, {name});')\
+                .format(code=support_code,
+                        name=name,
+                        implementation=implementation.replace('"', '\\"'),
```
Do we also need to escape the escapes ... so \ -> \\ before escaping quotes of course.

nikhilk replied:
Or why do we need to do this either ... the implementation isn't going in into a string.

gramster replied:
Correct, not needed.

nikhilk commented:
LGTM.
